transport/happyeyeballs.go

```diff
@@ -0,0 +1,237 @@
+		if err != nil {
+			err = fmt.Errorf("failed to lookup IPv4 addresses: %w", err)
+		}
+		lookup4Ch <- LookupResult{ips, err}
+		close(lookup4Ch)
+	}(lookup4Ch, host)
+
+	// DIAL ATTEMPTS SECTION
+	// We keep IPv4s and IPv6 separate and track the last one attempted so we can
+	// alternate the address family in the connection attempts.
+	var ip4s []netip.Addr
+	var ip6s []netip.Addr
+	var lastDialed netip.Addr
+	// Keep track of the lookup and dial errors separately. We prefer the dial errors
+	// when returning.
+	var lookupErr error
+	var dialErr error
+	// Channel to wait for before a new dial attempt. It starts
+	// with a closed channel that doesn't block because there's no
+	// wait initially.
+	var attemptDelayCh <-chan struct{} = newClosedChan()
+	type DialResult struct {
+		Conn StreamConn
+		Err  error
+	}
+	var dialCh = make(chan DialResult)
+
+	// Channel that triggers when a new connection can be made. Starts blocked (nil)
+	// because we need IPs first.
+	var readyToDialCh <-chan struct{} = nil
+	// We keep track of pending operations (lookups and IPs to dial) so we can stop when
+	// there's no more work to wait for.
+	for opsPending := 2; opsPending > 0; {
+		if len(ip6s) == 0 && len(ip4s) == 0 {
+			// No IPs. Keep dial disabled.
+			readyToDialCh = nil
+		} else {
+			// There are IPs to dial.
+			if !lastDialed.IsValid() && len(ip6s) == 0 && lookup6Ch != nil {
+				// Attempts haven't started and IPv6 lookup is not done yet. Set up Resolution Delay, as per
+				// https://datatracker.ietf.org/doc/html/rfc8305#section-8, if it hasn't been set up yet.
+				if readyToDialCh == nil {
+					resolutionDelayCtx, cancelResolutionDelay := context.WithTimeout(searchCtx, 50*time.Millisecond)
+					defer cancelResolutionDelay()
+					readyToDialCh = resolutionDelayCtx.Done()
+				}
+			} else {
+				// Wait for the previous attempt.
+				readyToDialCh = attemptDelayCh
+			}
+		}
+		select {
+		// Receive IPv6 results.
+		case lookupRes := <-lookup6Ch:
+			opsPending--
+			// Set to nil to make the read on lookup6Ch block and to signal IPv6 lookup is done.
+			lookup6Ch = nil
+			if lookupRes.Err != nil {
+				lookupErr = errors.Join(lookupRes.Err)
+				continue
+			}
+			opsPending += len(lookupRes.IPs)
+			// TODO: sort IPs as per https://datatracker.ietf.org/doc/html/rfc8305#section-4
+			ip6s = lookupRes.IPs
+
+		// Receive IPv4 results.
+		case lookupRes := <-lookup4Ch:
+			opsPending--
+			// Set to nil to make the read on lookup4Ch block and to signal IPv4 lookup is done.
+			lookup4Ch = nil
+			if lookupRes.Err != nil {
+				lookupErr = errors.Join(lookupRes.Err)
+				continue
+			}
+			opsPending += len(lookupRes.IPs)
+			// TODO: sort IPs as per https://datatracker.ietf.org/doc/html/rfc8305#section-4
+			ip4s = lookupRes.IPs
+
+		// Wait for new attempt done. Dial new IP address.
+		case <-readyToDialCh:
+			var toDial netip.Addr
+			if len(ip6s) == 0 || (lastDialed.Is6() && len(ip4s) > 0) {
```

Comment:
Shall we ensure `ip4s` or `ip6s` is not empty here? For example, when both `ip4s` and `ip6s` are empty, we will jump to the `toDial = ip4s[0]` part, are we confident that this will never happen?